routers/api/v1/repo/topic.go

```diff
@@ -0,0 +1,274 @@
+// license that can be found in the LICENSE file.
+
+package repo
+
+import (
+	"net/http"
+	"strings"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/log"
+	api "code.gitea.io/gitea/modules/structs"
+	"code.gitea.io/gitea/routers/api/v1/convert"
+)
+
+// ListTopics returns list of current topics for repo
+func ListTopics(ctx *context.APIContext) {
+	// swagger:operation GET /repos/{owner}/{repo}/topics repository repoListTopics
+	// ---
+	// summary: Get list of topics that a repository has
+	// produces:
+	//   - application/json
+	// parameters:
+	// - name: owner
+	//   in: path
+	//   description: owner of the repo
+	//   type: string
+	//   required: true
+	// - name: repo
+	//   in: path
+	//   description: name of the repo
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/TopicNames"
+
+	topics, err := models.FindTopics(&models.FindTopicOptions{
+		RepoID: ctx.Repo.Repository.ID,
+	})
+	if err != nil {
+		log.Error("ListTopics failed: %v", err)
+		ctx.JSON(500, map[string]interface{}{
+			"message": "ListTopics failed.",
+		})
+		return
+	}
+
+	topicNames := make([]*string, len(topics))
+	for i, topic := range topics {
+		topicNames[i] = &topic.Name
+	}
+	ctx.JSON(200, map[string]interface{}{
+		"topics": topicNames,
+	})
+}
+
+// UpdateTopics updates repo with a new set of topics
+func UpdateTopics(ctx *context.APIContext, form api.RepoTopicOptions) {
+	// swagger:operation PUT /repos/{owner}/{repo}/topics repository repoUpdateTopics
+	// ---
+	// summary: Replace list of topics for a repository
+	// produces:
+	//   - application/json
+	// parameters:
+	// - name: owner
+	//   in: path
+	//   description: owner of the repo
+	//   type: string
+	//   required: true
+	// - name: repo
+	//   in: path
+	//   description: name of the repo
+	//   type: string
+	//   required: true
+	// - name: body
+	//   in: body
+	//   schema:
+	//     "$ref": "#/definitions/RepoTopicOptions"
+	// responses:
+	//   "204":
+	//     "$ref": "#/responses/empty"
+
+	topicNames := form.Topics
+	validTopics, invalidTopics := models.SanitizeAndValidateTopics(topicNames)
+
+	if len(validTopics) > 25 {
+		ctx.JSON(422, map[string]interface{}{
+			"invalidTopics": nil,
+			"message":       "Exceeding maximum number of topics per repo",
+		})
+		return
+	}
+
+	if len(invalidTopics) > 0 {
+		ctx.JSON(422, map[string]interface{}{
+			"invalidTopics": invalidTopics,
+			"message":       "Topic names are invalid",
+		})
+		return
+	}
+
+	err := models.SaveTopics(ctx.Repo.Repository.ID, validTopics...)
+	if err != nil {
+		log.Error("SaveTopics failed: %v", err)
+		ctx.JSON(500, map[string]interface{}{
+			"message": "Save topics failed.",
+		})
+		return
+	}
+
+	ctx.Status(204)
+}
+
+// AddTopic adds a topic name to a repo
+func AddTopic(ctx *context.APIContext) {
+	// swagger:operation PUT /repos/{owner}/{repo}/topics/{topic} repository repoAddTopíc
```

Comment:
Generaly, we use POST to create an entry.

Reply:
I think this is a borderline case. Reading [this](https://restfulapi.net/rest-put-vs-post/), I find it match PUT more. You add the topic to the repo, not necessary create it, and notably the operation is idempotent (can only add topic once). Previous Gitea API endpoints use different:
- Add a collaborator to a repository: PUT
- Add a label to an issue: POST

Anyhow I guess it makes sense to use the same as for adding a label.

Reply:
_POST ​/repos​/{owner}​/{repo}​/issues​/{index}​/labels_ actually adds several labels (in body) and respond with what is added, so it makes more sense to use POST there. So I think it shall remain PUT here. If someone later wants to add an API to add several labels at the same time, that could be POST (although replace API should be enough).